### PR TITLE
SONARJAVA-1547 - SE: Support @Nonnull annotation on method invocations

### PR DESCRIPTION
**Pull request by DidierBesset**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java [hunk 1]
```diff
@@ -462,7 +463,27 @@ private void executeMethodInvocation(MethodInvocationTree mit) {
...
     logState(mit);
-    programState = programState.stackValue(constraintManager.createMethodSymbolicValue(mit, unstack.values));
+    final SymbolicValue resultValue = constraintManager.createMethodSymbolicValue(mit, unstack.values);
+    programState = programState.stackValue(resultValue);
+    if (isNonNullMethod(mit)) {
+      programState = programState.addConstraint(resultValue, ObjectConstraint.NOT_NULL);
+    }
+  }
+
+  private boolean isNonNullMethod(MethodInvocationTree mit) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: MAJOR Make "isNonNullMethod" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java [hunk 2]
```diff
@@ -462,7 +463,27 @@ private void executeMethodInvocation(MethodInvocationTree mit) {
...
+    IdentifierTree identifier;
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: This whole to get the identifier is useless. You can directly use `mit.symbol()` to get the symbol of the method.

File: java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java [hunk 3]
```diff
@@ -462,7 +463,27 @@ private void executeMethodInvocation(MethodInvocationTree mit) {
...
+    if (mit.methodSelect().is(Tree.Kind.IDENTIFIER)) {
+      identifier = (IdentifierTree) mit.methodSelect();
+    } else if (mit.methodSelect().is(Tree.Kind.MEMBER_SELECT)) {
+      identifier = ((MemberSelectExpressionTree) mit.methodSelect()).identifier();
+    } else {
+      return false;
+    }
+    if (identifier.symbol() instanceof JavaSymbol) {
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: You want to check if the method is _not unknown_ (`!symbol.isUnknown()`), not if it's a JavaSymbol!

File: java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java [hunk 4]
```diff
@@ -462,7 +463,20 @@ private void executeMethodInvocation(MethodInvocationTree mit) {
...
+    }
+  }
+
+  private static boolean isNonNullMethod(MethodInvocationTree mit) {
+    Symbol methodSymbol = mit.symbol();
+    if (methodSymbol.isUnknown()) {
+      return false;
+    }
+    JavaSymbol symbol = (JavaSymbol) methodSymbol;
+    return symbol.metadata().isAnnotatedWith("javax.annotation.Nonnull");
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: No need to cast to `JavaSymbol` to access the metadada.  In fact the method can even be refactored as : 
```
private static boolean isNonNullMethod(Symbol symbol) {
  return !symbol.isUnknown() && symbol.metadata().isAnnotatedWith("javax.annotation.Nonnull");
}
```

**Pull request merged**